[blackjack - Step1] Submit vittorio's mission

VittorioDeMarzi opened the pull request:
Pull Request: Step1 Implementation of BlackJack Game

Hi 🙂

Thanks in advance for taking the time to review our code and share your feedback. This pull request contain our implementation to the BlackJack game project. 

What wer did:
- Implemented the basic logic for a multiplayer BlackJack game.
- The game supports multiple players and a dealer, with separate rounds for each.
- Introduced a Statistics class to calculate the final results and print a summary.
- The GameManager handles the game setup and main loop.

What could be improved
- Currently, both Player and Dealer use the same class. Their behavior is distinguished using conditions.
- I'm considering introducing a common abstract class or interface (e.g., Participant) to separate responsibilities and avoid branching logic.

Questions I'd appreciate feedback on
- Do you think would it make more sense to extract a common interface or use an abstract class to share logic between them?
- Do you think the project structure and the separation into model, view, and controller packages follows the MVC pattern well, or are there improvements you’d suggest?
- Does GameManager take on too many responsibilities (setup, game logic, handling player vs dealer turns)?
- Is this method clear and maintainable? Or does the dealer vs player logic feel a bit forced?
- I’m currently passing InputView.askForCard() as a lambda to decouple input from game logic. Do you think this is a clean approach?

Looking forward to your thoughts and suggestions! 😊

VittorioDeMarzi committed:
docs(readme): first init

VittorioDeMarzi committed:
test: add a new test to implement Card class

VittorioDeMarzi committed:
test: update a test to implement Card class

VittorioDeMarzi committed:
feat: implement Card data class and Symbol

VittorioDeMarzi committed:
docs(readme): update the progress

VittorioDeMarzi committed:
test(domain): add a new test for wrapping class Cards

VittorioDeMarzi committed:
feat(domain): implement a mew wrapping class, Cards

VittorioDeMarzi committed:
test(domain): implement a mew test for CardDeck class.

VittorioDeMarzi committed:
test(domain): change name of test

VittorioDeMarzi committed:
feat(domain): implement CardDeck class and its initialization

VittorioDeMarzi committed:
feat(domain): implement operator plus in cards class

VittorioDeMarzi committed:
test(domain): add test for cardDeck shuffling

VittorioDeMarzi committed:
feat(domain): implement shuffle logic for cardDesk

VittorioDeMarzi committed:
refactor(domain): rename the class Cards to Hold

VittorioDeMarzi committed:
refactor(domain): change visibility of property to private and change name Cards class to Hold

VittorioDeMarzi committed:
test(domain): refactor name of test

VittorioDeMarzi committed:
test(domain): refactor test since priority of holds are private

VittorioDeMarzi committed:
refactor(domain): add methods in CardDeck to retrieve number of cards and list of cards

VittorioDeMarzi committed:
feat(domain): implement Hand class

VittorioDeMarzi committed:
test: refactor tests to moveCard to a Player

VittorioDeMarzi committed:
feat: implement Player class

VittorioDeMarzi committed:
feat: implement Person wrapper class for names

VittorioDeMarzi committed:
refactor(Hold): change target of moveCard to Player

VittorioDeMarzi committed:
docs(readme): update the progress

VittorioDeMarzi committed:
refactor: refactor packages to controller/model/view

VittorioDeMarzi committed:
docs(readme): add questions

VittorioDeMarzi committed:
feat(domain): add game controller and basic logic

VittorioDeMarzi committed:
test(domain): add new test for player factory class

VittorioDeMarzi committed:
feat(domain): implement PlayerFactory object

VittorioDeMarzi committed:
refactor(domain): add constructor to Player

VittorioDeMarzi committed:
test: refactor test to encapsulate the value

VittorioDeMarzi committed:
refactor: change visibility to property

VittorioDeMarzi committed:
refactor: implement method to create a dealer

VittorioDeMarzi committed:
refactor: increase readability of class Hand

VittorioDeMarzi committed:
test: add test to hit a card from the desk to a player

VittorioDeMarzi committed:
feat: implement hit method

VittorioDeMarzi committed:
refactor: remove unnecessary constructor and merge

VittorioDeMarzi committed:
refactor: adopt previews refactor

VittorioDeMarzi committed:
feat: implement GameManager class

VittorioDeMarzi committed:
test: add a new test to set up the table with GameManager

VittorioDeMarzi committed:
feat: implement GameManager and init logic

- dealer has a card
- each player have 2 cards

VittorioDeMarzi committed:
test(Hand): add tests for counting the points in one hand of a player

VittorioDeMarzi committed:
feat: implement logic for counting points in the hand of a player

VittorioDeMarzi committed:
feat: implement data class for a single player

VittorioDeMarzi committed:
refactor: adopt based to the previews feat

VittorioDeMarzi committed:
test: add test for get single player result

VittorioDeMarzi committed:
feat: get single player result and play with multiple players

VittorioDeMarzi committed:
refactor: adopt core logic to controller

VittorioDeMarzi committed:
refactor: add Rank to card

VittorioDeMarzi committed:
refactor: adapt to previews refactor

VittorioDeMarzi committed:
feat: add input validation for players names

VittorioDeMarzi committed:
feat: add Errors enum messages

VittorioDeMarzi committed:
refactor:remove card test not needed

VittorioDeMarzi committed:
refactor: adapt controller to logic of InputView

VittorioDeMarzi committed:
refactor: override toString

VittorioDeMarzi committed:
refactor: override toString and refactor get()

VittorioDeMarzi committed:
refactor: delete playerResult class

VittorioDeMarzi committed:
refactor: adapt controller and manager to previews refactor

VittorioDeMarzi committed:
feat: Output of first SetUp

VittorioDeMarzi committed:
refactor: ask for card In Input View

VittorioDeMarzi committed:
refactor: change logic of the game separating play with player or with dealer

VittorioDeMarzi committed:
refactor: Output View

VittorioDeMarzi committed:
refactor: adapt the controller to the previews refactor

VittorioDeMarzi committed:
refactor: add to player 2 properties, isBlackJack and isBust

VittorioDeMarzi committed:
feat: add and implement logic of statistics class

VittorioDeMarzi committed:
feat: print final result statistics

VittorioDeMarzi committed:
refactor: remove toString override

VittorioDeMarzi committed:
refactor: protect helper methods to private or internal: if needed to test

VittorioDeMarzi committed:
refactor: adopt prev commit

- change name of method inside GameManager

VittorioDeMarzi committed:
fix: remove not related class

VittorioDeMarzi committed:
refactor: add more logic to check user input

- check a result of split input as well

VittorioDeMarzi committed:
style: remove comments

VittorioDeMarzi committed:
refactor: adopt Statistics class

VittorioDeMarzi committed:
docs: update the progress

VittorioDeMarzi committed:
refactor: remove unused variable and constructor

Wordbe commented on src/main/kotlin/blackjack/model/GameManager.kt:
```diff
@@ -0,0 +1,64 @@
+class GameManager(private val dealer: Player, private val players: List<Player>) {
+    private val cardDeck = CardDeck()
+
+    fun setUp() {
```
If something must always be executed when GameManager is created, it might be a good idea to put it inside an init block! 😊

VittorioDeMarzi replied:
```
class GameManager(private val dealer: Player, private val players: List<Player>) {
    private val cardDeck = CardDeck()

    init {
        players.forEach { cardDeck.hit(it, 2) }
        cardDeck.hit(dealer)
    }
```

something like this?

Wordbe replied:
Yes exactly!

Wordbe commented on src/main/kotlin/blackjack/controller/BlackJackController.kt:
```diff
@@ -0,0 +1,28 @@
+            OutputView.printAllPlayers(listOf(dealer) + players)
+            gameManager.playGame(dealer, players) { InputView.askForCard() }
+
+            val statistics = Statistics(dealer, players)
```
It would be better to declare it on the line right after the code where it is used, keeping it close to where it's needed. 🙂

VittorioDeMarzi replied:
That's a simple but effective suggestion 🙂

Wordbe commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack.model
+
+data class Card(val suit: Suit, val rank: Rank) {
+    override fun toString(): String {
```
This method seems to be intended for the view, but it looks like the domain is depending on the view. How about separating them? 😊

Wordbe commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,25 @@
+            }
+
+        repeat(countOfAce) {
+            if (sum + 10 <= 21) sum += 10
```
Try defining the magic number as a constant!

Wordbe commented on src/main/kotlin/blackjack/model/Person.kt:
```diff
@@ -0,0 +1,4 @@
+package blackjack.model
+
+@JvmInline
+value class Person(val name: String)
```
The name Person suggests that the class could have many attributes beyond just name, so it might not be suitable as a value class. How about renaming it to something more specific? 🙂

Wordbe commented on src/main/kotlin/blackjack/model/Hold.kt:
```diff
@@ -0,0 +1,21 @@
+package blackjack.model
+
+class Hold(hold: Set<Card>) {
```
I'm curious what's the difference between Hand and Hold? 
If Hold is a value class wrapping a collection of Card objects, how about naming it Cards instead?

VittorioDeMarzi replied:
This is totally meaningful :D

Wordbe commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,25 @@
+
+    fun numberOfCards() = hold.cards.size
+
+    fun addCard(card: Card) = this.hold.add(card)
```
You could remove this here for consistency, since it's not used elsewhere. 😊

VittorioDeMarzi replied:
I am actually using it to hit the card to the player 😌

Wordbe replied:
I'm sorry I wasn't clear. I was referring to `this` 🙂

Wordbe commented on src/main/kotlin/blackjack/model/Statistics.kt:
```diff
@@ -0,0 +1,41 @@
+package blackjack.model
+
+class Statistics(private val dealer: Player, private val players: List<Player>) {
+    val totalResult
```
You might not need to create a member variable here!

Wordbe reviewed (changes requested):
Hello, Vito!
Nice to meet you.

Thank you for your thoughtful work on the Blackjack mission. The structure of your code looks pretty solid.

Please check the review comments, reflect on the parts that are feasible, and feel free to ask questions if anything is unclear! 😊

---
Q1.
Do you think would it make more sense to extract a common interface or use an abstract class to share logic between them?

It would be helpful to distinguish based on the presence or absence of state! 

Q2. Do you think the project structure and the separation into model, view, and controller packages follows the MVC pattern well, or are there improvements you’d suggest?

Yes, overall it looks good. However, it would be better to decouple the domain from the view dependency. 

Q3. Does GameManager take on too many responsibilities (setup, game logic, handling player vs dealer turns)?

Not too much right now, but it might be worth considering if things get a bit more complex. 

Q4. 
Is this method clear and maintainable? Or does the dealer vs player logic feel a bit forced?

It seems better to separate the dealer and players rather than managing them as a single player entity! 

Q5. 
I’m currently passing InputView.askForCard() as a lambda to decouple input from game logic. Do you think this is a clean approach?

Yes, there seem to be pros and cons to this pattern, and personally, I prefer simpler structures, so I don’t use it often. It can vary depending on the individual or the team!

---

Thank you!

VittorioDeMarzi committed:
refactor: add setUp carDeck in init instead of calling the function from the controller

VittorioDeMarzi committed:
refactor: declare statistics variable on the line right after the code where it is used

VittorioDeMarzi committed:
refactor: move responsibility ouf Card display to ResultView

VittorioDeMarzi committed:
refactor: define the magic number as a constant

VittorioDeMarzi committed:
refactor: rename Hold wrapper class to Cards

VittorioDeMarzi committed:
refactor: rename Hold wrapper class to Cards

VittorioDeMarzi committed:
test: adapt tests to previews refactors

VittorioDeMarzi committed:
feat: resolve dependency issue for OutputView and InputView

VittorioDeMarzi committed:
test: adapt tests to previews refactor

VittorioDeMarzi committed:
feat: remove totalResult and call the method directly instead of wrapping it in a property

VittorioDeMarzi committed:
feat: change name of Person simply to PlayerName

VittorioDeMarzi committed:
test: add test for dsl

VittorioDeMarzi committed:
feat: add implementation for dsl (name, company, skills)

VittorioDeMarzi committed:
test: add test for languages in dsl

VittorioDeMarzi committed:
feat: implement dsl languages block

VittorioDeMarzi committed:
feat: add dsl introduction to project

VittorioDeMarzi committed:
refactor: remove print function from OutputView

VittorioDeMarzi commented:
Hi Wordbe :)

Thanks a lot for the review and all your helpful feedback!

I’ve already addressed some of the points you mentioned, but I’ve kept the unified Player class for now instead of separating Dealer and Player. Before I go further with that refactor, I’d like to ask for your input.

Regarding your comment:
“It would be helpful to distinguish based on the presence or absence of state!”
I completely agree, and today in class we actually practiced applying the State Pattern (polymorphism) to the Blackjack project. It made me think it could be really interesting to integrate that approach into this project.
However, doing so would mean discarding most of the current structure and starting over from scratch to rebuild it around the pattern. It might take me a bit more time, do you think it’s worth trying now, or would it be better to first move on to Step 2?

Lastly, I also added a small DSL builder just to practice what we learned during the session. 🙂

Thanks again for your support  ☺️

Wordbe commented on src/main/kotlin/blackjack/dsl/PersonBuilder.kt:
```diff
@@ -0,0 +1,61 @@
+package blackjack.dsl
+
+data class Person(val name: String, val company: String?, val skills: Skills, val languages: Languages)
```
Great DSL 👍

Wordbe commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -0,0 +1,3 @@
+package blackjack.model
+
+data class Card(val suit: Suit, val rank: Rank)
```
Try limiting the number of Card instances that can be created infinitely. 🙂

VittorioDeMarzi replied:
I implemented the patter object pool in the Car Class. 
I think this is the right way for 
- Memory optimization: There are just 52 card in memory. No matter how many Desks we create. 
- Performance: We don't need to create new instances every time.
- Identity (Card1 === Card2 if they are equal)

Thanks for this feedback. This was really thoughtful. We've seen the Pool Pattern in Lotto with Jason but I didn't implemented yet :)

Wordbe commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    fun printStatistics(statistics: Statistics) {
+        println("\n## Final Results")
```
Combine multiple prints into a single string template and call println only once!

Wordbe reviewed (approved):
Hello Vito!
Since I believe you has reflected most of the feedback well in this step, let's go ahead and merge and wrap it up here. 😊

Here are a few suggestions for the next step:

- Please apply the remaining review comments I mentioned earlier at the beginning of the next step.

- As you mentioned, consider separating Player into Dealer and Player to reflect their distinct roles.

- Try managing each player's state using a State pattern.

Thank you for thoughtfully applying the reviews.
Great work!
